### PR TITLE
Mining charges, second try!

### PR DESCRIPTION
**Pull request by Henri215**
## What Does This PR Do

**Reopens https://github.com/ParadiseSS13/Paradise/pull/23515, the author had problems with the branch and couldnt solve the merge conflicts.**

Porting from yogs (with a small addition, taken directly from rupara) mining charges and related stuff.
Mining charges. They are bought from a mining vendor, obtained from a voucher and ordered from the RND. You can use it only on rocks, if not hacked.
There are 3 types of mining charges, they differ primarily in the radius of the explosion. After you attach the explosive to the rock, you got a 5 seconds before it detonates. If you stand close to them, you will be knocked down and get some ear- and eye damage, depending on the distance and power of charge. There is no real damage, so they are kinda safe
Lesser - 3 tiles (bought in vendor)
Standart - 5 tiles (bought in vendor)
Experimental - 8 tiles (made in RND)

https://private-user-images.githubusercontent.com/116907157/289315640-023091e6-b294-4b62-8238-29c6f834de61.mp4?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTEyODY2NDksIm5iZiI6MTcxMTI4NjM0OSwicGF0aCI6Ii8xMTY5MDcxNTcvMjg5MzE1NjQwLTAyMzA5MWU2LWIyOTQtNGI2Mi04MjM4LTI5YzZmODM0ZGU2MS5tcDQ_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwMzI0JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDMyNFQxMzE5MDlaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1iNTgzODM5N2M2NjNlMWM0NjcyM2I4OTMwOWRjMWI2MjVlOTMzZTBmZmJmNjgxNDAyNjdlN2ExY2Q1YjM0Y2E2JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.AWvmI3DbSpvji8ky62BdM8bUv8F7pOGzproIkL-CU84

Mining detonator. A special device that can link explosive charges to itself, so they aren't explode after you install them. You can track the quantity and state of explosives. After you active detonator, all installed explosives will detonate after a short time.

https://private-user-images.githubusercontent.com/116907157/289315807-99b8e527-2fec-4699-85ee-918f7f4c1d5e.mp4?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTEyODY2NDksIm5iZiI6MTcxMTI4NjM0OSwicGF0aCI6Ii8xMTY5MDcxNTcvMjg5MzE1ODA3LTk5YjhlNTI3LTJmZWMtNDY5OS04NWVlLTkxOGY3ZjRjMWQ1ZS5tcDQ_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwMzI0JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDMyNFQxMzE5MDlaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1lYzI3ZTM5OTMyYTllZDlkOTBhMWQ3MWQ2MWI2YjIyODkwMmI1YTc0NzMzNjIxNmViZTBjNDUyMGQ4YjUzYzYzJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.zuz7BydK-Ljz01Tz2tzjqyiEloQkULVC9myc9_XsvfE

Mining charge hacker.
New job-specific tool, cost 25TK. Has 6 charge. Allow user to reprogram mining charges to make them more lethal and allow user to install it everywhere they want. Reprogrammed charges are weaker than normal ones but still they can make holes in the station hull.
So explosion sizes:
Lesser - 0,0,2
Normal - 0,1,3
Experimental - 1,2,5

Hacked mining charges can be combined with a detonator to make nicer explosions as well.

https://private-user-images.githubusercontent.com/116907157/289316151-6e92cee6-7d42-4c2c-8d04-a50f9d08c5c2.mp4?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTEyODY2NDksIm5iZiI6MTcxMTI4NjM0OSwicGF0aCI6Ii8xMTY5MDcxNTcvMjg5MzE2MTUxLTZlOTJjZWU2LTdkNDItNGMyYy04ZDA0LWE1MGY5ZDA4YzVjMi5tcDQ_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwMzI0JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDMyNFQxMzE5MDlaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1hNGMzMjhhMzQ5MjY0ZTk0OTQ0OTgyZDE1NDEwZjQ4MDM4NGU1OTA1NTBkYjBjYzlkNGVkODg1MmVlZmY4ODZjJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.EVUK3qRmEugO0jfYpMWhcioCcnDZFU7LnOsL75Ox3Q8

Also, to make mining charges more attractive to user, all minerals will get the special +3 ore amount from ecah deposit mined by explosion.
And... added a new voucher bundle with 3 medium mining charges, 5 small mining charges and a detonator to play with.
## Why It's Good For The Game

New fun mechanic for lavaland, new way to mine ores, that people on ru-paradise (ex666, not that weirdos from ss220) liked very much. New fun way to destroy station.
## Images of changes

image

## Testing

Compiled and run. Destroyed some rocks, got more ore than usual.
Linked charges to a detonator, destroyed more rocks.
Hacked some charges, destroyed some walls.
Linked hacked charges to a detonator, destroyed some windows.
## Changelog
:cl: Henri215, NightDawnFox
add: Added mining charges, mining detonator and mining charge hacker.
tweak: Mineral rocks now drop +3 ores if destroyed by explosion.
/:cl:

### CODE REVIEW

**Comment by ppi13** [comment]
I do wonder how this will match the mining modsuit, considering it also has mining explosives already.

**Comment by Daylight2** [comment]
25TC seems really expensive, especially given that you have to still get the mining charges to use the tool on yourself. In comparison, a 5-pack of C4 is 20TC and seems pretty much identical in function, save for remote detonation. I'd even support making it an emag effect, although blowing into places after buying an emag might be a bit redundant.

**Comment by NightDawnFox** [comment]
> 25TC seems really expensive, especially given that you have to still get the mining charges to use the tool on yourself. In comparison, a 5-pack of C4 is 20TC and seems pretty much identical in function, save for remote detonation. I'd even support making it an emag effect, although blowing into places after buying an emag might be a bit redundant.

For 25 TC you get 6 charges, that definetly more powerful than normal c4 charges if you use medium or experimental charges, of course. Mega charges have an actual gib-radius, making it the most powerful plantable charge. And you can get 6 of them
image
There are charge explosion sizes. Only the cheapest charges are weaker than c4

File: code/game/objects/items/weapons/explosives.dm [hunk 1]
```diff
@@ -79,8 +80,9 @@
 		target = AM
 		loc = null
 
-		message_admins("[key_name_admin(user)]([ADMIN_QUE(user,"?")]) ([ADMIN_FLW(user,"FLW")]) planted [src.name] on [target.name] at ([target.x],[target.y],[target.z] - <A HREF='?_src_=holder;adminplayerobservecoodjump=1;X=[target.x];Y=[target.y];Z=[target.z]'>JMP</a>) with [det_time] second fuse",0,1)
-		log_game("[key_name(user)] planted [name] on [target.name] at ([target.x],[target.y],[target.z]) with [det_time] second fuse")
+		if(notify_admins)
+			message_admins("[key_name_admin(user)]([ADMIN_QUE(user,"?")]) ([ADMIN_FLW(user,"FLW")]) planted [src.name] on [target.name] at ([target.x],[target.y],[target.z] - <A HREF='?_src_=holder;adminplayerobservecoodjump=1;X=[target.x];Y=[target.y];Z=[target.z]'>JMP</a>) with [det_time] second fuse",0,1)
```
Comment (warriorstar-orion) [inline code/game/objects/items/weapons/explosives.dm:84]: while you're here

```suggestion
			message_admins("[key_name_admin(user)]([ADMIN_QUE(user,"?")]) ([ADMIN_FLW(user,"FLW")]) planted [src.name] on [target.name] at ([target.x],[target.y],[target.z] - <A HREF='?_src_=holder;adminplayerobservecoodjump=1;X=[target.x];Y=[target.y];Z=[target.z]'>JMP</a>) with [det_time] second fuse", 0, 1)
```

File: code/modules/mining/equipment/mining_charges.dm [hunk 2]
```diff
@@ -0,0 +1,213 @@
...
+	var/hacked = FALSE
+
+/obj/item/grenade/plastic/miningcharge/Initialize()
+	. = ..()
+	image_overlay = mutable_appearance(icon, "[icon_state]_active", ON_EDGED_TURF_LAYER)
+
+/obj/item/grenade/plastic/miningcharge/examine(mob/user)
+	. = ..()
+	if(hacked)
+		. += "Its wiring is haphazardly changed."
```
Comment (warriorstar-orion) [inline code/modules/mining/equipment/mining_charges.dm:24]: Should this have a span as well?

File: code/modules/mining/equipment/mining_charges.dm [hunk 3]
```diff
@@ -0,0 +1,213 @@
...
+		C.flash_eyes()
+		C.Weaken((boom_sizes[2] - distance) * 1 SECONDS) //1 second for how close you are to center if you're in range
+		C.AdjustDeaf((boom_sizes[3] - distance) * 5 SECONDS) //guaranteed deafness
+		var/obj/item/organ/internal/ears/ears = C.get_int_organ(/obj/item/organ/internal/ears)
+		if(istype(ears) && C.check_ear_prot() < HEARING_PROTECTION_MINOR) //headsets should be enough to avoid taking damage
+			ears.receive_damage((boom_sizes[3] - distance) * 2) //something like that i guess. Mega charge makes 12 damage to ears if nearby
+		to_chat(C, "<span class='userdanger'>You are knocked down by the power of the mining charge!</span>")
+	qdel(src)
+
+/obj/item/grenade/plastic/miningcharge/proc/explode() //c4 code
```
Comment (warriorstar-orion) [inline code/modules/mining/equipment/mining_charges.dm:96]: not a useful comment
```suggestion
/obj/item/grenade/plastic/miningcharge/proc/explode()
```

File: code/modules/mining/equipment/mining_charges.dm [hunk 4]
```diff
@@ -0,0 +1,213 @@
...
+	qdel(src)
+
+/obj/item/grenade/plastic/miningcharge/proc/override_safety()
+	hacked = TRUE
+	notify_admins = TRUE
+	boom_sizes[1] = round(boom_sizes[1] / 3)	//lesser - 0, normal - 0, mega - 1; c4 - 0
+	boom_sizes[2] = round(boom_sizes[2] / 3)	//lesser - 0, normal - 1, mega - 2; c4 - 0
+	boom_sizes[3] = round(boom_sizes[3] / 1.5)	//lesser - 2, normal - 3, mega - 5; c4 - 3
+
+/obj/item/grenade/plastic/miningcharge/deconstruct(disassembled = TRUE) //no gibbing a miner with pda bombs
```
Comment (warriorstar-orion) [inline code/modules/mining/equipment/mining_charges.dm:119]: PDA bombs?

File: code/modules/mining/equipment/mining_charges.dm (hunk 2 above)
Comment (Henri215) [reply code/modules/mining/equipment/mining_charges.dm]: Added on next commit

File: code/modules/mining/equipment/mining_charges.dm (hunk 4 above)
Comment (warriorstar-orion) [reply code/modules/mining/equipment/mining_charges.dm:119]: discussed on discord, this means that we're overriding the superclass so damage to these grenades from other sources don't cause them to go off.

Comment (Henri215) [reply code/modules/mining/equipment/mining_charges.dm:119]: Comment updated.

File: code/game/objects/items/weapons/explosives.dm [hunk 5]
```diff
@@ -79,8 +80,9 @@
 		target = AM
 		loc = null
 
-		message_admins("[key_name_admin(user)]([ADMIN_QUE(user,"?")]) ([ADMIN_FLW(user,"FLW")]) planted [src.name] on [target.name] at ([target.x],[target.y],[target.z] - <A HREF='?_src_=holder;adminplayerobservecoodjump=1;X=[target.x];Y=[target.y];Z=[target.z]'>JMP</a>) with [det_time] second fuse",0,1)
-		log_game("[key_name(user)] planted [name] on [target.name] at ([target.x],[target.y],[target.z]) with [det_time] second fuse")
+		if(notify_admins)
+			message_admins("[key_name_admin(user)]([ADMIN_QUE(user,"?")]) ([ADMIN_FLW(user,"FLW")]) planted [src.name] on [target.name] at ([target.x],[target.y],[target.z] - <A HREF='?_src_=holder;adminplayerobservecoodjump=1;X=[target.x];Y=[target.y];Z=[target.z]'>JMP</a>) with [det_time] second fuse", 0, 1)
```
Comment (Burzah) [inline code/game/objects/items/weapons/explosives.dm:84]:
```suggestion
			message_admins("[key_name_admin(user)]([ADMIN_QUE(user,"?")]) ([ADMIN_FLW(user,"FLW")]) planted [src.name] on [target.name] at ([target.x],[target.y],[target.z] - <a href='?_src_=holder;adminplayerobservecoodjump=1;X=[target.x];Y=[target.y];Z=[target.z]'>JMP</a>) with [det_time] second fuse", 0, 1)
```

File: code/modules/mining/equipment/mining_charges.dm [hunk 6]
```diff
@@ -0,0 +1,214 @@
...
+
+	/// list of all bombs connected to a detonator for a moment
+	var/list/bombs = list()
+
+/obj/item/detonator/examine(mob/user)
+	. = ..()
+	if(length(bombs))
+		. += "<span class='notice'>List of synched bombs:</span>"
+		for(var/obj/item/grenade/plastic/miningcharge/charge in bombs)
+			. += "<span class='notice'>[bicon(charge)] [charge]. Current status: [charge.installed ? "ready to detonate" : "ready to deploy"]."
```
Comment (Burzah) [inline code/modules/mining/equipment/mining_charges.dm:190]:
```suggestion
			. += "<span class='notice'>[bicon(charge)] [charge]. Current status: [charge.installed ? "ready to detonate" : "ready to deploy"].</span>"
```

**Review by Burzah — CHANGES_REQUESTED** [review]

**Review by warriorstar-orion — APPROVED** [review]
Whoops, sorry for the late turnaround on this. LGTM

File: code/game/objects/items/weapons/explosives.dm [hunk 7]
```diff
@@ -79,8 +80,9 @@
 		target = AM
 		loc = null
 
-		message_admins("[key_name_admin(user)]([ADMIN_QUE(user,"?")]) ([ADMIN_FLW(user,"FLW")]) planted [src.name] on [target.name] at ([target.x],[target.y],[target.z] - <A HREF='?_src_=holder;adminplayerobservecoodjump=1;X=[target.x];Y=[target.y];Z=[target.z]'>JMP</a>) with [det_time] second fuse",0,1)
-		log_game("[key_name(user)] planted [name] on [target.name] at ([target.x],[target.y],[target.z]) with [det_time] second fuse")
+		if(notify_admins)
+			message_admins("[key_name_admin(user)]([ADMIN_QUE(user,"?")]) ([ADMIN_FLW(user,"FLW")]) planted [src.name] on [target.name] at ([target.x],[target.y],[target.z] - <a href='?_src_=holder;adminplayerobservecoodjump=1;X=[target.x];Y=[target.y];Z=[target.z]'>JMP</a>) with [det_time] second fuse", 0, 1)
```
Comment (Contrabang) [inline code/game/objects/items/weapons/explosives.dm:84]:
```suggestion
			message_admins("[ADMIN_LOOKUPFLW(user)] planted [src.name] on [target.name] at ([target.x],[target.y],[target.z] - <a href='?_src_=holder;adminplayerobservecoodjump=1;X=[target.x];Y=[target.y];Z=[target.z]'>JMP</a>) with [det_time] second fuse", 0, 1)
```

**Review by Contrabang — CHANGES_REQUESTED** [review]
^review above, i pressed the wrong button

**Review by Burzah — APPROVED** [review]

File: code/game/objects/items/weapons/explosives.dm [hunk 8]
```diff
@@ -80,8 +80,9 @@
 		target = AM
 		loc = null
 
-		message_admins("[key_name_admin(user)]([ADMIN_QUE(user,"?")]) ([ADMIN_FLW(user,"FLW")]) planted [src.name] on [target.name] at ([target.x],[target.y],[target.z] - <A href='byond://?_src_=holder;adminplayerobservecoodjump=1;X=[target.x];Y=[target.y];Z=[target.z]'>JMP</a>) with [det_time] second fuse",0,1)
-		log_game("[key_name(user)] planted [name] on [target.name] at ([target.x],[target.y],[target.z]) with [det_time] second fuse")
+		if(notify_admins)
+			message_admins("[ADMIN_LOOKUPFLW(user)] planted [src.name] on [target.name] at ([target.x],[target.y],[target.z] - <A href='byond://?_src_=holder;adminplayerobservecoodjump=1;X=[target.x];Y=[target.y];Z=[target.z]'>JMP</a>) with [det_time] second fuse", 0, 1)
```
Comment (Burzah) [inline code/game/objects/items/weapons/explosives.dm:84]:
```suggestion
			message_admins("[ADMIN_LOOKUPFLW(user)] planted [src.name] on [target.name] at ([target.x],[target.y],[target.z] - <a href='byond://?_src_=holder;adminplayerobservecoodjump=1;X=[target.x];Y=[target.y];Z=[target.z]'>JMP</a>) with [det_time] second fuse", 0, 1)
```

**Review by Burzah — CHANGES_REQUESTED** [review]

File: code/modules/mining/equipment/mining_charges.dm [hunk 9]
```diff
@@ -0,0 +1,214 @@
...
+	if(do_after(user, (2.5 SECONDS) * toolspeed, target = AM))
+		if(!user.unEquip(src))
+			return
+		target = AM
+		loc = null
+		if(hacked)
+			message_admins("[ADMIN_LOOKUPFLW(user)] planted [name] on [target.name] at [ADMIN_COORDJMP(target)]")
+			log_game("planted [name] on [target.name] at [COORD(target)]", user)
+		installed = TRUE
+		target.overlays += image_overlay
```
Comment (DGamerL) [inline code/modules/mining/equipment/mining_charges.dm:54]:
```suggestion
	if(!do_after(user, (2.5 SECONDS) * toolspeed, target = AM))
		return
	if(!user.unEquip(src))
		return
	target = AM
	forceMove(AM)
	if(hacked)
		message_admins("[ADMIN_LOOKUPFLW(user)] planted [name] on [target.name] at [ADMIN_COORDJMP(target)]")
		log_game("planted [name] on [target.name] at [COORD(target)]", user)
	installed = TRUE
	target.overlays += image_overlay
```

File: code/modules/mining/equipment/mining_charges.dm [hunk 10]
```diff
@@ -0,0 +1,214 @@
...
+		return
+	var/obj/item/detonator/detonator = I
+	if(!(src in detonator.bombs) && !timer_off)
+		detonator.bombs += src
+		timer_off = TRUE
+		to_chat(user, "<span class='notice'>You synchronize [src] to [I].</span>")
+		playsound(src, 'sound/machines/twobeep.ogg', 50)
+		detonator.update_icon()
+	else
+		to_chat(user, "<span class='warning'>[src] was already synchronized to a existing detonator!</span>")
```
Comment (DGamerL) [inline code/modules/mining/equipment/mining_charges.dm:67]:
```suggestion
	if((src in detonator.bombs) || timer_off)
		to_chat(user, "<span class='warning'>[src] was already synchronized to a existing detonator!</span>")
		return
	detonator.bombs += src
	timer_off = TRUE
	to_chat(user, "<span class='notice'>You synchronize [src] to [I].</span>")
	playsound(src, 'sound/machines/twobeep.ogg', 50)
	detonator.update_icon()
```

File: code/modules/mining/equipment/mining_charges.dm [hunk 11]
```diff
@@ -0,0 +1,214 @@
...
+
+/obj/item/detonator/attack_self(mob/user)
+	playsound(src, 'sound/items/detonator.ogg', 40)
+	if(length(bombs))
+		to_chat(user, "<span class='notice'>Activating explosives...</span>")
+		for(var/obj/item/grenade/plastic/miningcharge/charge in bombs)
+			if(QDELETED(charge))
+				bombs -= charge
+				update_icon() //if the last bomb was qdeleted, detonator icon should change after activating
+				return
```
Comment (DGamerL) [inline code/modules/mining/equipment/mining_charges.dm:207]:
```suggestion
				continue
```

**Pull request merged**